### PR TITLE
Distributed and Batch AUC

### PR DESCRIPTION
**Pull request by seiriosPlus**

### CODE REVIEW

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 1]
```diff
@@ -866,6 +871,76 @@ def _update_dist_lookup_table_vars(self, param_list, grad_list,
                 ]
         return param_list, grad_list
 
+    def _append_distributed_metric(self):
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py:966]: can add a new transpiler? input a distributed program, and return a program with distributed metric?

Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py:966]: What is the meaning of distributed program? The training program or the parameter server program?

Comment (seiriosPlus) [reply python/paddle/fluid/transpiler/distribute_transpiler.py:966]: I think you are right, the distributed metrics are independent and multitudinous,   I will separate it from the current distributed program.

**Comment by gongweibao** [comment]
Please add an ISSUE of it first!

**Comment by seiriosPlus** [comment]
The Issue is: https://github.com/PaddlePaddle/Paddle/issues/13205

File: paddle/fluid/operators/auc_op.cc [hunk 2]
```diff
@@ -36,11 +36,16 @@ class AucOp : public framework::OperatorWithKernel {
                       "Out and Label should have same height.");
 
     int num_pred_buckets = ctx->Attrs().Get<int>("num_thresholds") + 1;
```
Comment (jacquesqiao) [inline paddle/fluid/operators/auc_op.cc]: num_pred_buckets => num_thresholds
命名保持一致？

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 3]
```diff
@@ -874,6 +879,76 @@ def _update_dist_lookup_table_vars(self, param_list, grad_list,
...
+            self.has_distributed_metrics = False
+            return
+
+        self.has_distributed_metrics = True
+
+        for op in self.distributed_metric_ops:
+            if op.type == "auc":
+                self._append_distributed_auc(op)
+
+    def _append_distributed_auc(self, auc_op):
```
Comment (typhoonzero) [inline python/paddle/fluid/transpiler/distribute_transpiler.py:984]: can we not use a transpiler? like using some code like:

server: 
```python
with fluid.layers.ListenAndServe():
    fluid.layers.auc()
```

clients:
```python
...
fluid.layers.send()
```

File: paddle/fluid/operators/auc_op.cc [hunk 4]
```diff
@@ -36,11 +36,16 @@ class AucOp : public framework::OperatorWithKernel {
...
+
+    PADDLE_ENFORCE_GE(num_pred_buckets, 1, "num_thresholds must larger than 1");
+    PADDLE_ENFORCE_GE(slide_steps, 0, "slide_steps must be natural number");
 
     ctx->SetOutputDim("AUC", {1});
-    ctx->SetOutputDim("BatchAUC", {1});
-    ctx->SetOutputDim("StatPosOut", {num_pred_buckets});
-    ctx->SetOutputDim("StatNegOut", {num_pred_buckets});
+
+    slide_steps = slide_steps == 0 ? 1 : slide_steps;
```
Comment (jacquesqiao) [inline paddle/fluid/operators/auc_op.cc:46]: 前边已经有一个check了
```cpp
PADDLE_ENFORCE_GE(slide_steps, 0, "slide_steps must be natural number");
```

**Comment by seiriosPlus** [comment]
Batch AUC has been merged in another PR.
Distributed AUC will be rewritten after RPC SEND/RECV multi vars.

**Pull request closed without merging**